% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Individual_Analysis_PheWAS.R
\name{Individual_Analysis_PheWAS}
\alias{Individual_Analysis_PheWAS}
\title{Individual-variant PheWAS analysis using score test}
\usage{
Individual_Analysis_PheWAS(
  chr,
  start_loc,
  end_loc,
  genofile,
  obj_nullmodel_list,
  mac_cutoff = 20,
  subset_variants_num = 5000,
  QC_label = "annotation/filter",
  variant_type = c("variant", "SNV", "Indel"),
  geno_missing_imputation = c("mean", "minor"),
  tol = .Machine$double.eps^0.25,
  max_iter = 1000,
  SPA_p_filter = TRUE,
  p_filter_cutoff = 0.05
)
}
\arguments{
\item{chr}{chromosome.}

\item{start_loc}{starting location (position) of the genetic region for each individual variant to be analyzed using score test.}

\item{end_loc}{ending location (position) of the genetic region for each individual variant to be analyzed using score test.}

\item{genofile}{an object of opened annotated GDS (aGDS) file.}

\item{obj_nullmodel_list}{a list of objects from fitting the null models, which are either the output from \code{fit_nullmodel} function in the \code{STAARpipeline} package,
or the output from \code{fitNullModel} function in the \code{GENESIS} package and transformed using the \cr \code{genesis2staar_nullmodel} function in the \code{STAARpipeline} package.}

\item{mac_cutoff}{the cutoff of minimum minor allele count in
defining individual variants (default = 20).}

\item{subset_variants_num}{the number of variants to run per subset for each time (default = 5e3).}

\item{QC_label}{channel name of the QC label in the GDS/aGDS file (default = "annotation/filter").}

\item{variant_type}{type of variant included in the analysis. Choices include "variant", "SNV", or "Indel" (default = "variant").}

\item{geno_missing_imputation}{method of handling missing genotypes. Either "mean" or "minor" (default = "mean").}

\item{tol}{a positive number specifying tolerance, the difference threshold for parameter
estimates in saddlepoint approximation algorithm below which iterations should be stopped (default = ".Machine$double.eps^0.25").}

\item{max_iter}{a positive integer specifying the maximum number of iterations for applying the saddlepoint approximation algorithm (default = "1000").}

\item{SPA_p_filter}{logical: are only the variants with a score-test-based p-value smaller than a pre-specified threshold use the SPA method to recalculate the p-value, only used for imbalanced case-control setting (default = TRUE).}

\item{p_filter_cutoff}{threshold for the p-value recalculation using the SPA method, only used for imbalanced case-control setting (default = 0.05)}
}
\value{
A list of data frames containing the score test p-value and the estimated effect size of the minor allele for each individual variant in the given genetic region for each phenotype.
The first 4 columns of each data frame correspond to chromosome (CHR), position (POS), reference allele (REF), and alternative allele (ALT).
}
\description{
The \code{Individual_Analysis_PheWAS} function takes in chromosome, starting location, ending location,
the object of opened annotated GDS file, and the list of objects from fitting the null models to analyze the association between a series of
quantitative/dichotomous phenotypes (including imbalanced case-control design) and each individual variant in a genetic region by using score test.
For multiple phenotype analysis (\code{obj_nullmodel$n.pheno > 1}),
the results correspond to multi-trait score test p-values by leveraging
the correlation structure between multiple phenotypes.
}
\references{
Chen, H., et al. (2016). Control for population structure and relatedness for binary traits
in genetic association studies via logistic mixed models. \emph{The American Journal of Human Genetics}, \emph{98}(4), 653-666.
(\href{https://doi.org/10.1016/j.ajhg.2016.02.012}{pub})

Li, Z., Li, X., et al. (2022). A framework for detecting
noncoding rare-variant associations of large-scale whole-genome sequencing
studies. \emph{Nature Methods}, \emph{19}(12), 1599-1611.
(\href{https://doi.org/10.1038/s41592-022-01640-x}{pub})
}
